common/flatpak-dir.c

```diff
@@ -4748,13 +4748,97 @@ export_desktop_file (const char   *app,
   return ret;
 }
 
+static GStrv
+enumerate_owned_dbus_names (GKeyFile  *metakey,
+                            GError   **error)
+{
+  g_autoptr(GError) local_error = NULL;
+  g_autoptr(GPtrArray) owned_names = g_ptr_array_new_with_free_func (g_free);
+  g_auto(GStrv) keys = g_key_file_get_keys (metakey,
+                                            FLATPAK_METADATA_GROUP_SESSION_BUS_POLICY,
+                                            NULL,
+                                            &local_error);
+  GStrv iter = keys;
+
+  if (keys == NULL)
+    {
+      /* No group, return an empty strv */
+      if (g_error_matches (local_error,
+                           G_KEY_FILE_ERROR,
+                           G_KEY_FILE_ERROR_GROUP_NOT_FOUND))
+        {
+          g_ptr_array_add (owned_names, NULL);
+          return (GStrv) g_ptr_array_free (owned_names, FALSE);
+        }
+
+      g_propagate_error (error, g_steal_pointer (&local_error));
+      return NULL;
+    }
+
+  for (; *iter != NULL; ++iter)
+    {
+      g_autofree char *value = g_key_file_get_string (metakey,
+                                                      FLATPAK_METADATA_GROUP_SESSION_BUS_POLICY,
+                                                      *iter,
+                                                      error);
+
+      if (value == NULL)
+        return NULL;
+
+      if (g_strcmp0 (value, "own") == 0)
+        g_ptr_array_add (owned_names, g_strdup (*iter));
+    }
+
+  g_ptr_array_add (owned_names, NULL);
+  return (GStrv) g_ptr_array_free (g_steal_pointer (&owned_names), FALSE);
+}
+
+typedef gboolean (*MultiplePrefixesComparisonFunc) (const char         *path,
+                                                    const char * const *prefixes);
+
+static GStrv
+get_permissible_prefixes (GKeyFile                        *metakey,
+                          const char                      *source_path,
+                          const char                      *app_id,
+                          MultiplePrefixesComparisonFunc  *out_match_prefixes_func,
```

Comment:
Why do you need this function returned? No valid app id will ever end with `.*`, so you could always use the wildcarding one.

Reply:
So the returned function here is the policy to use in detecting which files should be exported. In the dbus services case we want to use the wildcarding one since we either want to export the dbus service files that match the --own-name specification completely or ones that match a wildcarded specification. In all other cases, we still want to use prefix based matching.

Reply:
Yes, but the "allowable prefixes" will never include a string ending with ".*" in the regular case, so it is safe to always use the wildcard matching method.

Reply:
My point here was more that the wildcard matching method checks for an *exact name match* on the file if not wildcard is specified, whereas for the other paths only need to have a prefix match for them to be exported.